tests/wasmer-argus/src/argus/packages.rs

```diff
@@ -0,0 +1,217 @@
+        p.set_message("starting to fetch packages..".to_string());
+
+        while let Some(pkgs) = call.next().await {
+            let pkgs = match pkgs {
+                Ok(pkgs) => pkgs,
+                Err(e) => {
+                    error!("failed to fetch packages: {e}");
+                    p.finish_and_clear();
+                    anyhow::bail!("failed to fetch packages: {e}")
+                }
+            };
+            p.set_message(format!("fetched {} packages", count));
+            count += pkgs.len();
+
+            for pkg in pkgs {
+                if self.to_test(&pkg).await {
+                    if let Err(e) = s.send(pkg) {
+                        error!("failed to send packages: {e}");
+                        p.finish_and_clear();
+                        anyhow::bail!("failed to send packages: {e}")
+                    };
+                }
+            }
+        }
+
+        p.finish_with_message(format!("fetched {count} packages"));
+        info!("finished fetching packages: fetched {count} packages, closing channel");
+        drop(s);
+        Ok(())
+    }
+
+    #[tracing::instrument(skip(p))]
+    pub(crate) async fn download_package<'a>(
+        test_id: u64,
+        path: &'a PathBuf,
+        url: &'a Url,
+        p: &'a ProgressBar,
+    ) -> anyhow::Result<()> {
+        info!("downloading package from {} to file {:?}", url, path);
+        static APP_USER_AGENT: &str =
+            concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"),);
+
+        if !path.exists() {
+            tokio::fs::create_dir_all(path).await?;
+        } else if path.exists() && !path.is_dir() {
+            anyhow::bail!("path {:?} exists, but it is not a directory!", path)
+        }
+
+        let client = Client::builder().user_agent(APP_USER_AGENT).build()?;
+
+        let download_size = {
+            let resp = client.head(url.as_str()).send().await?;
+            if resp.status().is_success() {
+                resp.headers()
+                    .get(header::CONTENT_LENGTH)
+                    .and_then(|ct_len| ct_len.to_str().ok())
+                    .and_then(|ct_len| ct_len.parse().ok())
+                    .unwrap_or(0) // Fallback to 0
+            } else {
+                anyhow::bail!(
+                    "Couldn't fetch head from URL {}. Error: {:?}",
+                    url,
+                    resp.status()
+                )
+            }
+        };
+
+        let request = client.get(url.as_str());
+
+        p.set_length(download_size);
+
+        p.set_style(
+            ProgressStyle::default_bar()
+                .template(&format!(
+                    "[{test_id}] [{{bar:40.cyan/blue}}] {{bytes}}/{{total_bytes}} - {{msg}}"
+                ))
+                .unwrap()
+                .progress_chars("#>-"),
+        );
+
+        p.set_message(format!("downloading from {url}"));
+
+        let mut outfile = match File::create(&path.join("package.webc")) {
```

Comment:
Using a sync file in an async context, this should be a tokio File.

Reply:
Edited in [9f066cc](https://github.com/wasmerio/wasmer/pull/4476/commits/9f066cccd38ac80756f25d4c6d2bab1be2bf6110)